0021-zarf-values/README.md

```diff
@@ -0,0 +1,291 @@
+
+This feature also could open up Zarf packages to being less declarative - especially if a package author opens up security-critical Helm values in their charts.  This caveat should be clearly documented as a concern which should also recommend a policy engine be used to enforce security-critical values within the cluster itself.
+
+Because we will be using more `interface{}` types, we should also look into the security implications of this feature and ensure that this is well tested and we utilize some of Helm's existing protections against `nil` maps and other potential security issues with this feature.
+
+## Design Details
+
+The new `values.files` field would be added to the `ZarfPackageConfig` schema and would accept a list of local values files or URLs, matching the existing functionality of the `valuesFiles` key under `charts`.  This key would also follow the same composability logic as the `valuesFiles` key with additional parent values files merging with and overriding any common keys from children imports.  Since this is a global field, values would be merged regardless of component similar to how variables work today.  The `values.schema` field would simply replace the child version if it were non-empty in the parent, similar to the `namespace` or `releaseName` fields in `charts` today.
+
+Each of the referenced values files would be included inside the created Zarf package and stored inside of the tarball or at an OCI path so that they would travel with the package and be available to the user on deploy.  On deploy, the values (from the defaults in the `ZarfPackageConfig` and the set values files) would be merged with the set values overriding the defaults, and the resulting values would be passed to the Helm chart via the `values` field under a given `charts` entry.  Any actions which contained a go template (i.e. `{{ .Values.component.resources | toYaml }}`) would be templated prior to execution. This templating would work in all fields within an action definition including `cmd` and `wait` actions.
+
+This feature would also implement a `setValues` field for actions that would act as a replacement for the existing `setVariables` field.  This field would take a values path and would set that path to the output of the command.  This field would also have `type` defined on it, though instead of `file`, it would take `string`, `json`, or `yaml` and then handle the standard out of the command according to that format.
+
+Zarf Values would also be available to onCreate and onRemove actions if they existed in `package.create.values` or `package.remove.values` or were set with `-f`/`--values`.  These would template actions and be able to use `setValues` but would not map to Helm charts.
```

Comment:
I do like the idea of opening up variables to `onRemove` in this way. This was requested in https://github.com/zarf-dev/zarf/issues/1801, while this issue is closed, it's a reasonable thing to expect Zarf to do.

I am leaning towards not allowing go templates during `onCreate` actions. Variables are never loaded on create so they can't currently be used in actions, even as defaults. I worry that running a command like `zarf package create -f my-values.yaml` may lead a user to believe that their values file is now loaded into their package to be used at deploy time. Technically that's possible, but it would lessen the declarative nature of `zarf package create` more than is necessary imo.

Reply:
I agree actually - been too long since I have even used `onCreate` sets that I forgot they were completely different - will update.